File: tests/rowColumnStats.py
#!/usr/bin/env python2

#
# LSST Data Management System
# Copyright 2008, 2009, 2010 LSST Corporation.
#
# This product includes software developed by the
# LSST Project (http://www.lsst.org/).
#
# This program is free software: you can redistribute it and/or modify
# it under the terms of the GNU General Public License as published by
# the Free Software Foundation, either version 3 of the License, or
# (at your option) any later version.
#
# This program is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
# GNU General Public License for more details.
#
# You should have received a copy of the LSST License Statement and
# the GNU General Public License along with this program.  If not,
# see <http://www.lsstcorp.org/LegalNotices/>.
#

# -*- python -*-
"""
Tests for statisticsStack row/column statistics

Run with:
   ./rowColumnStats.py
or
   python
   >>> import rowColumnStats; rowColumnStats.run()
"""
from __future__ import division
from builtins import range

##########################
# rowColumnStats.py
# Steve Bickerton
# An python test to check the row/column statistics from statisticsStack

import unittest
import lsst.afw.image as afwImage
import lsst.afw.math as afwMath
import lsst.afw.geom as afwGeom
import lsst.utils.tests as utilsTests


class RowColumnStatisticsTestCase(unittest.TestCase):

    def setUp(self):

        # fill an image with a gradient
        self.n      = 8
        self.img    = afwImage.ImageF(afwGeom.Extent2I(self.n, self.n), 0)

        # these are the known answers for comparison
        def nVector(n, v):
            return [v for i in range(n)]
        self.column = nVector(self.n, 0.0)
        self.row    = nVector(self.n, 0.0)
        self.colPlus  = nVector(self.n, 0.0)
        self.colMinus = nVector(self.n, 0.0)
        self.colMult  = nVector(self.n, 0.0)
        self.colDiv   = nVector(self.n, 0.0)
        self.rowPlus  = nVector(self.n, 0.0)
        self.rowMinus = nVector(self.n, 0.0)
        self.rowMult  = nVector(self.n, 0.0)
        self.rowDiv   = nVector(self.n, 0.0)

        # set the values in the image, and keep track of the stats to verify things
        for y in range(self.n):
            for x in range(self.n):
                val = 1.0*x + 2.0*y
                self.img.set(x, y, val)
                self.column[y] += val
                self.row[x]    += val

        for i in range(self.n):
            self.row[i]    /= self.n
            self.column[i] /= self.n
            self.colPlus[i] = self.img.get(0, i) + self.column[i]

        # get stats on the columns and rows
        self.imgProjectCol = afwMath.statisticsStack(self.img, afwMath.MEAN, 'x')
        self.imgProjectRow = afwMath.statisticsStack(self.img, afwMath.MEAN, 'y')


    def tearDown(self):
        del self.img
        del self.imgProjectCol
        del self.imgProjectRow

    def testColumnStats(self):
        """Test the column statistics """
        for i in range(self.n):
            self.assertEqual(self.imgProjectCol.get(0, i)[0], self.column[i])

    def testRowStats(self):
        """Test the row statistics """
        for i in range(self.n):
            self.assertEqual(self.imgProjectRow.get(i, 0)[0], self.row[i])


    def testColumnOperators(self):
        """ Test operator overloading on columns """

        columnSlice = afwImage.ImageSliceF(self.imgProjectCol.getImage())

        imgAdd = self.img + columnSlice
        imgAdd2 = columnSlice + self.img
        imgSub = self.img - columnSlice
        imgMul = self.img * columnSlice
        imgMul2 = columnSlice * self.img
        imgDiv = self.img / columnSlice

        for i in range(self.n):
            self.assertAlmostEqual(imgAdd.get(0, i), self.img.get(0, i) + columnSlice.get(0, i))
            self.assertAlmostEqual(imgAdd2.get(0, i), imgAdd.get(0, i))
            self.assertAlmostEqual(imgSub.get(0, i), self.img.get(0, i) - columnSlice.get(0, i))
            self.assertAlmostEqual(imgMul.get(0, i), self.img.get(0, i) * columnSlice.get(0, i))
            self.assertAlmostEqual(imgMul2.get(0, i), imgMul.get(0, i))
            self.assertAlmostEqual(imgDiv.get(0, i), self.img.get(0, i) / columnSlice.get(0, i))


#################################################################
# Test suite boiler plate
#################################################################
def suite():
    """Returns a suite containing all the test cases in this module."""

    utilsTests.init()

    suites = []
    suites += unittest.makeSuite(RowColumnStatisticsTestCase)
    suites += unittest.makeSuite(utilsTests.MemoryTestCase)
    return unittest.TestSuite(suites)

def run(shouldExit = False):
    """Run the tests"""
    utilsTests.run(suite(), shouldExit)

if __name__ == "__main__":
    run(True)

File: tests/rowColumnStats.cc
// -*- lsst-c++ -*-

/*
 * LSST Data Management System
 * Copyright 2008, 2009, 2010 LSST Corporation.
 *
 * This product includes software developed by the
 * LSST Project (http://www.lsst.org/).
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the LSST License Statement and
 * the GNU General Public License along with this program.  If not,
 * see <http://www.lsstcorp.org/LegalNotices/>.
 */

/*
 * An test executible which calls the statisticsStack function
 */
#include <iostream>
#include <memory>

#define BOOST_TEST_DYN_LINK
#define BOOST_TEST_MODULE rowColumnStatistics

#pragma clang diagnostic push
#pragma clang diagnostic ignored "-Wunused-variable"
#include "boost/test/unit_test.hpp"
#pragma clang diagnostic pop
#include "boost/test/tools/floating_point_comparison.hpp"

#include "lsst/geom.h"
#include "lsst/afw/image/Image.h"
#include "lsst/afw/math/Stack.h"

namespace image = lsst::afw::image;
namespace math = lsst::afw::math;

using ImageF = image::Image<float>;
using MImageF = image::MaskedImage<float>;

BOOST_AUTO_TEST_CASE(RowColumnStats) { /* parasoft-suppress  LsstDm-3-2a LsstDm-3-4a LsstDm-4-6 LsstDm-5-25
                                          "Boost non-Std" */

    int const n = 8;

    // fill an image with a gradient
    std::vector<float> column(n, 0.0);
    std::vector<float> row(n, 0.0);
    std::shared_ptr<ImageF> img = std::make_shared<ImageF>(lsst::geom::Extent2I(n, n), 0);
    for (int y = 0; y < img->getHeight(); ++y) {
        int x = 0;
        for (ImageF::x_iterator ptr = img->row_begin(y), end = img->row_end(y); ptr != end; ++ptr, ++x) {
            *ptr = 1.0 * x + 2.0 * y;
            column[y] += *ptr;
            row[x] += *ptr;
            if (y == n - 1) {
                row[x] /= n;
            }
        }
        column[y] /= n;
    }

    // collapse with a MEAN over 'x' (ie. avg all columns to one), then 'y' (avg all rows to one)
    std::shared_ptr<MImageF> imgProjectCol = math::statisticsStack(*img, math::MEAN, 'x');
    std::shared_ptr<MImageF> imgProjectRow = math::statisticsStack(*img, math::MEAN, 'y');

    MImageF::x_iterator rPtr = imgProjectRow->row_begin(0);
    MImageF::y_iterator cPtr = imgProjectCol->col_begin(0);
    for (int i = 0; i < n; ++i, ++rPtr, ++cPtr) {
        BOOST_CHECK_EQUAL(cPtr.image(), column[i]);
        BOOST_CHECK_EQUAL(rPtr.image(), row[i]);
    }
}
